core/src/main/java/org/apache/iceberg/UpdateManifests.java

```diff
@@ -0,0 +1,151 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.OutputFile;
+
+abstract class UpdateManifests<ThisT> extends SnapshotProducer<ThisT> {
+
+  private static final String KEPT_MANIFESTS_COUNT = "manifests-kept";
+  private static final String NEW_MANIFESTS_COUNT = "manifests-created";
+  private static final String REPLACED_MANIFESTS_COUNT = "manifests-replaced";
+  private static final String PROCESSED_ENTRY_COUNT = "entries-processed";
+
+  private final List<ManifestFile> keptManifests = Collections.synchronizedList(new ArrayList<>());
+  private final List<ManifestFile> newManifests = Collections.synchronizedList(new ArrayList<>());
+  private final Set<ManifestFile> replacedManifests = Collections.synchronizedSet(new HashSet<>());
+
+  private final AtomicInteger manifestSuffix = new AtomicInteger(0);
+  private final AtomicLong entryCount = new AtomicLong(0);
+
+  private final SnapshotSummary.Builder summaryBuilder = SnapshotSummary.builder();
+
+  UpdateManifests(TableOperations ops) {
+    super(ops);
+  }
+
+  @Override
+  protected String operation() {
+    return DataOperations.REPLACE;
+  }
+
+  @Override
+  public ThisT set(String property, String value) {
+    summaryBuilder.set(property, value);
+    return self();
+  }
+
+  protected SnapshotSummary.Builder summaryBuilder() {
+    return summaryBuilder;
+  }
+
+  @Override
+  protected Map<String, String> summary() {
+    summaryBuilder.set(KEPT_MANIFESTS_COUNT, Integer.toString(keptManifests.size()));
+    summaryBuilder.set(NEW_MANIFESTS_COUNT, Integer.toString(newManifests.size()));
+    summaryBuilder.set(REPLACED_MANIFESTS_COUNT, Integer.toString(replacedManifests.size()));
+    summaryBuilder.set(PROCESSED_ENTRY_COUNT, Long.toString(entryCount.get()));
+    return summaryBuilder.build();
+  }
+
+  @Override
+  protected List<ManifestFile> apply(TableMetadata base) {
+    updateManifests(base);
+    validateFileCounts();
+
+    // put new manifests at the beginning
+    return ImmutableList.<ManifestFile>builder()
+        .addAll(newManifests)
+        .addAll(keptManifests)
+        .build();
+  }
+
+  protected abstract void updateManifests(TableMetadata base);
+
+  private void validateFileCounts() {
+    int newFileCount = newManifests.stream().mapToInt(ManifestFile::existingFilesCount).sum();
```

Comment:
We have to discuss this part. In theory, `existingFilesCount` can be null but I am not sure in which case that will happen. Instead of assuming it is always not null, we can skip the file count check if we detect null in any of the replaced or new manifests.

Reply:
I think it is reasonable to require that the ManifestFile instances passed to this class have the counts set. This should reject updates if counts are null so that this validation can assume they are present.